File: frontend/lib/api/caption-presets.ts
import { API_BASE_URL } from "@/lib/config";
import type {
  CaptionPreset,
  CaptionPresetExportPayload,
  CaptionPresetImportPayload,
  CaptionPresetListResponse,
  CaptionPresetStyle,
} from "@/lib/caption-presets";

export type ApiError = {
  message: string;
  status?: number;
};

async function parseError(response: Response): Promise<ApiError> {
  try {
    const payload = (await response.json()) as { detail?: string | Array<{ msg?: string }> };
    if (typeof payload.detail === "string") {
      return { message: payload.detail, status: response.status };
    }
    if (Array.isArray(payload.detail) && payload.detail[0]?.msg) {
      return { message: payload.detail[0].msg ?? "Request failed.", status: response.status };
    }
  } catch {
    // ignore parse errors
  }
  return { message: response.statusText || "Request failed.", status: response.status };
}

export async function fetchCaptionPresets(): Promise<CaptionPresetListResponse> {
  const response = await fetch(`${API_BASE_URL}/api/caption-presets`);
  if (!response.ok) {
    throw await parseError(response);
  }
  return response.json();
}

export async function createCaptionPreset(request: {
  name: string;
  style: CaptionPresetStyle;
}): Promise<CaptionPreset> {
  const response = await fetch(`${API_BASE_URL}/api/caption-presets`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(request),
  });
  if (!response.ok) {
    throw await parseError(response);
  }
  return response.json();
}

export async function updateCaptionPreset(
  presetId: string,
  request: {
    name?: string;
    style?: CaptionPresetStyle;
    is_default?: boolean;
  },
): Promise<CaptionPreset> {
  const response = await fetch(`${API_BASE_URL}/api/caption-presets/${presetId}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(request),
  });
  if (!response.ok) {
    throw await parseError(response);
  }
  return response.json();
}

export async function deleteCaptionPreset(presetId: string): Promise<void> {
  const response = await fetch(`${API_BASE_URL}/api/caption-presets/${presetId}`, {
    method: "DELETE",
  });
  if (!response.ok) {
    throw await parseError(response);
  }
}

export async function duplicateCaptionPreset(presetId: string): Promise<CaptionPreset> {
  const response = await fetch(`${API_BASE_URL}/api/caption-presets/${presetId}/duplicate`, {
    method: "POST",
  });
  if (!response.ok) {
    throw await parseError(response);
  }
  return response.json();
}

export async function exportCaptionPreset(presetId: string): Promise<CaptionPresetExportPayload> {
  const response = await fetch(`${API_BASE_URL}/api/caption-presets/${presetId}/export`);
  if (!response.ok) {
    throw await parseError(response);
  }
  return response.json();
}

export async function importCaptionPresets(
  payload: CaptionPresetImportPayload,
): Promise<{ imported: CaptionPreset[] }> {
  const response = await fetch(`${API_BASE_URL}/api/caption-presets/import`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!response.ok) {
    throw await parseError(response);
  }
  return response.json();
}

File: backend/app/api/routes/caption_presets.py
from __future__ import annotations

from fastapi import APIRouter, HTTPException, Response
from pydantic import ValidationError

from app.models.caption_preset import (
    CaptionPresetExportPayload,
    CaptionPresetImportPayload,
    CaptionPresetImportResponse,
    CaptionPresetListResponse,
    CaptionPresetResponse,
    CreateCaptionPresetRequest,
    UpdateCaptionPresetRequest,
)
from app.services.caption_presets import (
    CaptionPresetConflictError,
    CaptionPresetNotFoundError,
    CaptionPresetValidationError,
    create_caption_preset,
    delete_caption_preset,
    duplicate_caption_preset,
    export_caption_preset,
    get_caption_preset,
    import_caption_presets,
    list_caption_presets,
    update_caption_preset,
)

router = APIRouter(prefix="/api/caption-presets", tags=["caption-presets"])


@router.get("", response_model=CaptionPresetListResponse)
def list_presets() -> CaptionPresetListResponse:
    return list_caption_presets()


@router.post("", response_model=CaptionPresetResponse, status_code=201)
def create_preset(request: CreateCaptionPresetRequest) -> CaptionPresetResponse:
    try:
        return create_caption_preset(request)
    except CaptionPresetValidationError as exc:
        raise HTTPException(status_code=422, detail=exc.message) from exc
    except ValidationError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc


@router.post("/import", response_model=CaptionPresetImportResponse)
def import_presets(payload: CaptionPresetImportPayload) -> CaptionPresetImportResponse:
    try:
        imported = import_caption_presets(payload)
        return CaptionPresetImportResponse(imported=imported)
    except CaptionPresetValidationError as exc:
        raise HTTPException(status_code=422, detail=exc.message) from exc
    except ValidationError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc


@router.get("/{preset_id}", response_model=CaptionPresetResponse)
def get_preset(preset_id: str) -> CaptionPresetResponse:
    try:
        return get_caption_preset(preset_id)
    except CaptionPresetNotFoundError as exc:
        raise HTTPException(status_code=404, detail=exc.message) from exc


@router.patch("/{preset_id}", response_model=CaptionPresetResponse)
def patch_preset(preset_id: str, request: UpdateCaptionPresetRequest) -> CaptionPresetResponse:
    try:
        return update_caption_preset(preset_id, request)
    except CaptionPresetNotFoundError as exc:
        raise HTTPException(status_code=404, detail=exc.message) from exc
    except CaptionPresetConflictError as exc:
        raise HTTPException(status_code=409, detail=exc.message) from exc
    except CaptionPresetValidationError as exc:
        raise HTTPException(status_code=422, detail=exc.message) from exc
    except ValidationError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc


@router.delete("/{preset_id}", status_code=204, response_class=Response)
def remove_preset(preset_id: str) -> Response:
    try:
        delete_caption_preset(preset_id)
    except CaptionPresetNotFoundError as exc:
        raise HTTPException(status_code=404, detail=exc.message) from exc
    except CaptionPresetConflictError as exc:
        raise HTTPException(status_code=409, detail=exc.message) from exc
    return Response(status_code=204)


@router.post("/{preset_id}/duplicate", response_model=CaptionPresetResponse, status_code=201)
def duplicate_preset(preset_id: str) -> CaptionPresetResponse:
    try:
        return duplicate_caption_preset(preset_id)
    except CaptionPresetNotFoundError as exc:
        raise HTTPException(status_code=404, detail=exc.message) from exc
    except CaptionPresetValidationError as exc:
        raise HTTPException(status_code=422, detail=exc.message) from exc


@router.get("/{preset_id}/export", response_model=CaptionPresetExportPayload)
def export_preset(preset_id: str) -> CaptionPresetExportPayload:
    try:
        return export_caption_preset(preset_id)
    except CaptionPresetNotFoundError as exc:
        raise HTTPException(status_code=404, detail=exc.message) from exc
